### PR TITLE
Hash routing

### PR DESCRIPTION
**Pull request by timwis**
This pull request leverages the recent work of @nlocnila in yoshuawuyts/sheet-router#18 to add hash routing support. The current code works, but could be optimized. Wanted to get feedback on the direction first.

Note that to use this, users must use `app.start(null, {history: false, href: false})`. Otherwise `href` hijacks the anchor tags.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -204,7 +213,16 @@ function appInit (opts) {
   // enable catching <href a=""></href> links
   // enable HTML5 history API
   if (opts.href !== false) pushLocationSub(href)
-  if (opts.history !== false) pushLocationSub(history)
+  if (opts.history !== false) {
+    pushLocationSub(history)
+  // otherwise enable hash history
+  } else {
+    pushLocationSub(function (navigate) {
```
Comment (timwis) [inline index.js]: The only way I can think to shorten this like the other two is to use function composition, which would require a `compose` function or importing one from a library.

File: index.js [hunk 2]
```diff
@@ -187,9 +189,16 @@ function choo () {
 // initial application state model
 // obj -> obj
 function appInit (opts) {
+  var initialLocation
+  if (opts.history !== false) {
```
Comment (timwis) [inline index.js]: Some redundancy here as `opts.history !== false` is performed below too.

Comment (yoshuawuyts) [inline index.js]: perhaps a ternary if it fits?

File: index.js [hunk 3]
```diff
@@ -204,7 +213,16 @@ function appInit (opts) {
   // enable catching <href a=""></href> links
   // enable HTML5 history API
   if (opts.href !== false) pushLocationSub(href)
-  if (opts.history !== false) pushLocationSub(history)
+  if (opts.history !== false) {
+    pushLocationSub(history)
+  // otherwise enable hash history
+  } else {
```
Comment (yoshuawuyts) [inline index.js]: perhaps have `opts.hash === true` as the trigger? can imagine there's situations where neither is desirable to enable

File: index.js (hunk 1 above)
Comment (yoshuawuyts) [inline index.js]: nah, it's probably ok (':

**Comment by yoshuawuyts** [comment]
yeah, if this works then sure!

**Comment by timwis** [comment]
Updated per your feedback. If there aren't any other implications I'm overlooking, I can add this to the docs and the PR will be ready to merge.

File: index.js [hunk 4]
```diff
@@ -190,9 +192,13 @@ function choo () {
...
+    ? hashMatch(document.location.hash)
+    : document.location.href
+
   const model = {
     namespace: 'app',
-    state: { location: document.location.href },
+    state: { location: initialLocation },
     subscriptions: [],
     reducers: {
       // handle href links
```
Comment (timwis) [inline index.js]: Not sure the `.replace(/#.*/, '')` is necessary anymore with hash-match in use, though I'm not sure what purpose this served originally if it was meant to be used with browser history.

Comment (yoshuawuyts) [inline index.js]: Oh yeah, that was to support `<a href="#"></a>` style links; not sure it's needed anymore, but eh - let's not remove it quite yet :sparkles:

**Comment by yoshuawuyts** [comment]
Oh damn, last thing for this: we need to update the docs

**Comment by timwis** [comment]
Doc update added 📜

**Comment by yoshuawuyts** [comment]
Merged in `v2.2.0` - thanks a lot @timwis :sparkles:

**Pull request merged**